projects/newprofile_ping_nightly_validation.kp/knowledge.md

```diff
@@ -0,0 +1,518 @@
+      'meta/X-PingSender-Version': u'1.0',
+      'meta/creationTimestamp': 1.496505394245e+18,
+      'payload/reason': u'shutdown'},
+     {'clientId': 'XXX',
+      'environment/profile/creationDate': 17295,
+      'id': 'XXX',
+      'meta/Date': u'Sat, 03 Jun 2017 22:00:26 GMT',
+      'meta/Timestamp': 1496530831045648896L,
+      'meta/X-PingSender-Version': None,
+      'meta/creationTimestamp': 1.496486568754e+18,
+      'payload/reason': u'shutdown'},
+     {'clientId': 'XXX',
+      'environment/profile/creationDate': 17295,
+      'id': 'XXX',
+      'meta/Date': u'Sat, 03 Jun 2017 22:30:26 GMT',
+      'meta/Timestamp': 1496532637876130304L,
+      'meta/X-PingSender-Version': None,
+      'meta/creationTimestamp': 1.496529026725e+18,
+      'payload/reason': u'startup'},
+     {'clientId': 'XXX',
+      'environment/profile/creationDate': 17295,
+      'id': 'XXX',
+      'meta/Date': u'Sun, 04 Jun 2017 20:22:30 GMT',
+      'meta/Timestamp': 1496607753696764928L,
+      'meta/X-PingSender-Version': None,
+      'meta/creationTimestamp': 1.496607377349e+18,
+      'payload/reason': u'shutdown'}]
+
+
+
+Interesting! Looks like many of these pings are missing the pingsender header from the request, meaning:
+
+- that they are not being sent from the pingsender, even though they are generated at shutdown;
+- that we generated a new-profile ping at shutdown, but failed to mark it as 'generated', and so we received more than one with a different document it.
+
+This leads to other questions:
+
+- How often do we send new-profile pings at shutdown, fail, and then send them again without the pingsender?
+- Does that correlate with the duplicates?
+
+
+```python
+newprofile_shutdown_from_bad_clients =\
+    subset.filter(lambda p: p.get("payload/reason") == 'shutdown')\
+          .filter(lambda p: p.get("clientId") in misbehaving_clients)
+
+newprofile_shutdown_from_good_clients =\
+    subset.filter(lambda p: p.get("payload/reason") == 'shutdown')\
+          .filter(lambda p: p.get("clientId") not in misbehaving_clients)
+```
+
+```python
+newprofile_shutdown_from_bad_clients\
+    .map(lambda p: p.get("meta/X-PingSender-Version")).countByValue()
+```
+
+
+
+
+    defaultdict(int, {None: 10, u'1.0': 32})
+
+
+
+
+```python
+newprofile_shutdown_from_good_clients\
+    .map(lambda p: p.get("meta/X-PingSender-Version")).countByValue()
+```
+
+
+
+
+    defaultdict(int, {None: 528, u'1.0': 2336})
+
+
+
+Sending `new-profile` pings with reason `shutdown` seems to happen half of the times with misbehaving clients and about 10% of the times with well behaving clients. The pingsender doesn't seem to be the issue here: if we generate a ping at shutdown and try to send it with the pingsender, and fail, then it's normal for Firefox to pick it back up and send it. As long as we don't generate a new, different, `new-profile` ping for the same client. And this is something that well behaving clients don't do, so that 10% should be safe.
```

Comment:
Whoops. Looks like we're receiving about ~75% of the pings with the pingsender (~81% from well behaved clients). Changed this too with the correct numbers: the conclusion stays the same, looks like the pingsender in itself isn't the problem there, as most of the pings are being received from the pingsender with both good and bad behaving clients.